### ISSUE TITLE
baseCleanMount does not unmount all filesystems

### ISSUE THREAD

**Issue by vinyanalista** [openSUSE/kiwi#501]
I was trying to build an appliance with the following KIWI configuration:

https://github.com/kamarada/kiwi-config-Kamarada/tree/f3f46ac736e4af181ffc20cf80178da9acb53347

It is an openSUSE 42.1 LiveDVD with KDE, using as starting point the template `livecd-kde` found in the package [kiwi-config-openSUSE](https://software.opensuse.org/package/kiwi-config-openSUSE) for openSUSE 13.2 and the SUSE Studio great `create_appliance.sh`.

To build it, I was using a VirtualBox virtual machine with openSUSE 42.1 Milestone 1 for 64 bit PC, installed from Network, and [KIWI v7.03.23](https://github.com/openSUSE/kiwi/tree/v7.03.23), cloned from GitHub. KIWI configuration files were located at `/hd/kiwi-config-Kamarada/` (I needed to add another HD to my virtual machine, because the default HD of 8 GB went out of space).

My `config.sh` file follows the example in `man KIWI::config.sh`:

```sh
#======================================
# Functions...
#−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−
test −f /.kconfig && . /.kconfig
test −f /.profile && . /.profile

#======================================
# Greeting...
#−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−
echo "Configure image: [$kiwi_iname]..."

#======================================
# Mount system filesystems
#−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−
baseMount

#======================================
# Call configuration code/functions
#−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−
...

#======================================
# Umount kernel filesystems
#−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−
baseCleanMount

#======================================
# Exit safely
#−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−
exit 0
```

But it was always returning the same error when it was ran during the prepare step:

```
Umount of /hd/kiwi-config-Kamarada/image/build/image-root/sys failed: umount: /hd/kiwi-config-Kamarada/image/build/image-root/sys: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
```

In order to at least complete the prepare step, I replaced `baseCleanMount` by `baseCleanMount || true`. Then, `config.sh` exited successfully, although ignoring that error. The create step, then, failed just at the beginning, showing me the following error message:

```
It appears there are processes which are holding onto mounts created by a previous run

...

This could cause kiwi to fail, or even worse, if you are building on tmpfs, it could totally freeze or crash your machine. Please first clean up these old mounts, or export 'KIWI_IGNORE_OLD_MOUNTS=yes' and re-run 
Closing session with ecode: 1
KIWI exited with error(s)
Complete logfile at: /hd/kiwi-config-Kamarada/image/build/image-root.log
```

I ran `export KIWI_IGNORE_OLD_MOUNTS=yes` and then `kiwi --create` and it built the appliance successfully. Now I'm going to test if the appliance has any configuration mistakes on my part, but it seems like KIWI has done its job. The appliance successfully booted and showed the KDE Plasma desktop with the SUSEgreeter screen.

I reproduced the same sequence of errors using another VirtualBox virtual machine with openSUSE 42.1 Beta 1 for 64 bit PC, installed from Network, and [KIWI v7.03.25](https://github.com/openSUSE/kiwi/tree/v7.03.25), cloned from GitHub.

Am I doing something wrong, or is this a KIWI bug?

**Comment by vinyanalista** [comment]
Also, I would like to note that I need to change `dir="$(dirname $0)"` to `dir="$(pwd)"` in the script `create_appliance.sh` for KIWI to work. I haven't committed that change yet because I'm not sure if there is a problem with my setup. I cloned KIWI from GitHub and then added its location to the `PATH` environment variable. Then, `kiwi --version` answered accordingly.

I suspect that there is a problem with my setup because I never needed to do that while using KIWI on openSUSE 13.2, the version provided in the official repositories.

**Comment by v-fox** [comment]
It's probably the same thing I've encountered and was thinking about reporting for a while:
lack of unmount for "/sys/kernel/security". Which I worked around by having:
```
umount -f -l /sys/kernel/security
baseCleanMount
```
In config.sh and launching kiwi via script that forces it just for good measure.

**Comment by vinyanalista** [comment]
These days, I have been successfully building my appliance, I don't know what I changed in its configuration in order to be able to build it... the last time I tried to build my appliance, I tried to use less memory (I set the virtual machine to use only 512MB), KIWI ran out of memory. Today I tried again, this time with 1.5GB, KIWI crashed again, like before. Here is what I did:

1) I just started my VirtualBox virtual machine (this time with openSUSE Leap 42.1 stable, everything up to date) with that same [KIWI v7.03.23](https://github.com/openSUSE/kiwi/tree/v7.03.23) in text only mode (passing the argument `2` to the kernel via the bootloader menu)
2) Logged in as normal user
3) Switched to the appliance folder
4) Ran `sudo su`
5) Called `sh create_appliance.sh`, which failed.

Here is probably the most relevant part of the `image-root.log` file (I was not able to attach it entirely because GitHub didn't allow me):

```
Nov-09 09:23:47 <1> : Removing packages marked for deletion:
Nov-09 09:23:47 <1> : --> catdoc
Nov-09 09:23:47 <1> : --> kernel-debug
Nov-09 09:23:47 <1> : Removing packages...
Nov-09 09:23:47 <1> : EXEC [chroot "/home/kamarada/kiwi-config-Kamarada/image/build/image-root" /usr/bin/rpmdb --initdb &>/dev/null]
Nov-09 09:23:47 <1> : EXEC [chroot "/home/kamarada/kiwi-config-Kamarada/image/build/image-root" ldconfig 2>&1]
Nov-09 09:23:47 <1> : EXEC [chmod 755 /home/kamarada/kiwi-config-Kamarada/image/build/image-root/screenrc.smart]
+ export PBL_SKIP_BOOT_TEST=1
+ PBL_SKIP_BOOT_TEST=1
+ export LANG=C
+ LANG=C
+ export LC_ALL=C
+ LC_ALL=C
+ trap clean INT TERM
+ chroot /home/kamarada/kiwi-config-Kamarada/image/build/image-root mount -t proc proc /proc
++ grep -v 'is not installed'
++ chroot /home/kamarada/kiwi-config-Kamarada/image/build/image-root rpm -q catdoc kernel-debug
+ final=catdoc-0.94.4-5.2.x86_64
+ '[' '!' -z catdoc-0.94.4-5.2.x86_64 ']'
+ SPID=1728
+ wait 1728
+ chroot /home/kamarada/kiwi-config-Kamarada/image/build/image-root rpm -e --nodeps --allmatches --noscripts catdoc-0.94.4-5.2.x86_64
warning: file /usr/share/doc/packages/catdoc/TODO: remove failed: No such file or directory
warning: file /usr/share/doc/packages/catdoc/README: remove failed: No such file or directory
warning: file /usr/share/doc/packages/catdoc/NEWS: remove failed: No such file or directory
warning: file /usr/share/doc/packages/catdoc/CREDITS: remove failed: No such file or directory
warning: file /usr/share/doc/packages/catdoc/COPYING: remove failed: No such file or directory
warning: file /usr/share/doc/packages/catdoc: remove failed: No such file or directory
+ ECODE=0
+ echo 0
+ chroot /home/kamarada/kiwi-config-Kamarada/image/build/image-root umount /proc
Nov-09 09:23:48 <1> : EXEC [rm -f /home/kamarada/kiwi-config-Kamarada/image/build/image-root/screenrc.smart*]
Nov-09 09:23:48 <1> : EXEC [rm -f /home/kamarada/kiwi-config-Kamarada/image/build/image-root/screenrc.ctrls]
Nov-09 09:23:48 <1> : EXEC [rm -f /home/kamarada/kiwi-config-Kamarada/image/build/image-root/screenrc.err]
Nov-09 09:23:48 <1> : EXEC [diff -q /etc/resolv.conf /home/kamarada/kiwi-config-Kamarada/image/build/image-root/etc/resolv.conf]
Nov-09 09:23:48 <1> : Cleanup temporary copy of resolv.conf
Nov-09 09:23:48 <1> : EXEC [rm -f /home/kamarada/kiwi-config-Kamarada/image/build/image-root/etc/resolv.conf]
Nov-09 09:23:48 <1> : EXEC [diff -q /etc/sysconfig/proxy /home/kamarada/kiwi-config-Kamarada/image/build/image-root/etc/sysconfig/proxy]
Nov-09 09:23:48 <1> : Cleanup temporary copy of sysconfig/proxy
Nov-09 09:23:48 <1> : EXEC [cp /home/kamarada/kiwi-config-Kamarada/image/build/image-root/var/adm/fillup-templates/sysconfig.proxy /home/kamarada/kiwi-config-Kamarada/image/build/image-root/etc/sysconfig/proxy]
Nov-09 09:23:48 <1> : Release package manager lock
Nov-09 09:23:48 <1> : EXEC [rm -f /var/lock/kiwi-init.lock]
Nov-09 09:23:48 <1> : Release package manager lock
Nov-09 09:23:48 <1> : EXEC [rm -f /var/lock/kiwi-init.lock]
Nov-09 09:23:48 <1> : EXEC [test -d "/home/kamarada/kiwi-config-Kamarada/image/build/image-root//var/cache/kiwi"]
Nov-09 09:23:48 <1> : Umounting path: /home/kamarada/kiwi-config-Kamarada/image/build/image-root//var/cache/kiwi
Nov-09 09:23:48 <1> : EXEC [umount "/home/kamarada/kiwi-config-Kamarada/image/build/image-root//var/cache/kiwi" 2>&1]
Nov-09 09:23:48 <1> : EXEC [test -d "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/proc/sys/fs/binfmt_misc"]
Nov-09 09:23:48 <1> : Warning: "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/proc/sys/fs/binfmt_misc" not a directory or not existing
Nov-09 09:23:48 <1> : EXEC [test -d "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/dev/pts"]
Nov-09 09:23:48 <1> : Umounting path: /home/kamarada/kiwi-config-Kamarada/image/build/image-root/dev/pts
Nov-09 09:23:48 <1> : EXEC [umount "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/dev/pts" 2>&1]
Nov-09 09:23:48 <1> : EXEC [test -d "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/sys"]
Nov-09 09:23:48 <1> : Umounting path: /home/kamarada/kiwi-config-Kamarada/image/build/image-root/sys
Nov-09 09:23:48 <1> : EXEC [umount "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/sys" 2>&1]
Nov-09 09:23:48 <2> : Umount of /home/kamarada/kiwi-config-Kamarada/image/build/image-root/sys failed: umount: /home/kamarada/kiwi-config-Kamarada/image/build/image-root/sys: target is busy
        (In some cases useful info about processes that
         use the device is found by lsof(8) or fuser(1).)
Nov-09 09:23:48 <1> : EXEC [test -d "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/var/run/dbus"]
Nov-09 09:23:48 <1> : Umounting path: /home/kamarada/kiwi-config-Kamarada/image/build/image-root/var/run/dbus
Nov-09 09:23:48 <1> : EXEC [umount "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/var/run/dbus" 2>&1]
Nov-09 09:23:48 <1> : EXEC [test -d "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/dev"]
Nov-09 09:23:48 <1> : Umounting path: /home/kamarada/kiwi-config-Kamarada/image/build/image-root/dev
Nov-09 09:23:48 <1> : EXEC [umount "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/dev" 2>&1]
Nov-09 09:23:48 <1> : EXEC [test -d "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/proc"]
Nov-09 09:23:48 <1> : Umounting path: /home/kamarada/kiwi-config-Kamarada/image/build/image-root/proc
Nov-09 09:23:48 <1> : EXEC [umount "/home/kamarada/kiwi-config-Kamarada/image/build/image-root/proc" 2>&1]
Nov-09 09:23:48 <1> : EXEC [/usr/bin/uname -m]
Nov-09 09:23:49 <1> : EXEC [/usr/bin/uname -m]
Nov-09 09:23:49 <1> : Description provides no MD5 hash, check
Nov-09 09:23:49 <1> : Reading image description [Create]...
Nov-09 09:23:49 <1> : EXEC [/usr/bin/uname -m]
Nov-09 09:23:49 <3> : It appears there are processes which are holding onto mounts created by a previous run 
systemd(1)-+-ModemManager(736)-+-{gdbus}(796)
           |                   `-{gmain}(794)
           |-VBoxService(819)-+-{automount}(826)
           |                  |-{control}(820)
           |                  |-{cpuhotplug}(823)
           |                  |-{memballoon}(824)
           |                  |-{timesync}(821)
           |                  |-{vminfo}(822)
           |                  `-{vmstats}(825)
           |-auditd(549)---{auditd}(556)
           |-avahi-daemon(732)
           |-cron(1495)
           |-cupsd(1307)
           |-dbus-daemon(729)
           |-dmeventd(366)
           |-haveged(387)
           |-login(773)---bash(2917)---sudo(2971)---su(2972)---bash(2973)---sh(3020)---kiwi(3029)---kiwi.pl(3031)---pstree(1779)
           |-master(1478)-+-pickup(1483)
           |              `-qmgr(1484)
           |-nscd(748)-+-{nscd}(752)
           |           |-{nscd}(753)
           |           |-{nscd}(754)
           |           |-{nscd}(755)
           |           |-{nscd}(756)
           |           |-{nscd}(757)
           |           |-{nscd}(758)
           |           |-{nscd}(759)
           |           |-{nscd}(760)
           |           `-{nscd}(761)
           |-ntpd(1393)
           |-polkitd(797)-+-{JS GC Helper}(831)
           |              |-{JS Sour~ Thread}(832)
           |              |-{gdbus}(830)
           |              |-{gmain}(829)
           |              `-{runaway-killer-}(833)
           |-systemd(2914)---(sd-pam)(2916)
           |-systemd-journal(361)
           |-systemd-logind(762)
           |-systemd-udevd(378)
           |-wickedd(873)
           |-wickedd-auto4(871)
           |-wickedd-dhcp4(870)
           |-wickedd-dhcp6(869)
           |-wickedd-nanny(874)
           `-wpa_supplicant(737)

Nov-09 09:23:49 <3> : This could cause kiwi to fail, or even worse, if you are building on tmpfs, it could totally freeze or crash your machine. Please first clean up these old mounts, or export 'KIWI_IGNORE_OLD_MOUNTS=yes' and re-run 
Nov-09 09:23:49 <1> : Closing session with ecode: 1
Nov-09 09:23:50 <3> : KIWI exited with error(s)
Nov-09 09:23:50 <1> : EXEC [mv /home/kamarada/kiwi-config-Kamarada/image/build/image-root.3031.screenrc.log /home/kamarada/kiwi-config-Kamarada/image/build/image-root.log 2>&1]
Nov-09 09:23:51 <1> : Complete logfile at: /home/kamarada/kiwi-config-Kamarada/image/build/image-root.log
Nov-09 09:23:51 <1> : Release package manager lock
Nov-09 09:23:51 <1> : EXEC [rm -f /var/lock/kiwi-init.lock]
Nov-09 09:23:51 <1> : Release package manager lock
Nov-09 09:23:51 <1> : EXEC [rm -f /var/lock/kiwi-init.lock]
Nov-09 09:23:51 <1> : EXEC [rm -f /var/cache/kiwi/zypper/zypper.conf.3031 /var/cache/kiwi/zypper/zypp.conf.3031]
```

The appliance I tried to build can be found here:

https://github.com/kamarada/kiwi-config-Kamarada/tree/c4cc8926baf54d37b8026d9cd8924fbbf480a4e6

It seems like always when it fails, it complains about not being able to unmount `/sys`. I'm going to try some things: maybe switch to the KIWI stable version from the openSUSE Leap 42.1 repository, `export 'KIWI_IGNORE_OLD_MOUNTS=yes'`, @v-fox's suggestion... if I successfully build this appliance I tell you what I did.

Just an update to a previous comment of mine: I'm now definitely using `dir="$(pwd)"` on my `create_appliance.sh` script, as you can see at [that link](https://github.com/kamarada/kiwi-config-Kamarada/tree/c4cc8926baf54d37b8026d9cd8924fbbf480a4e6).

**Comment by schaefi** [comment]
Hmm, I think this "mount" check in kiwi is simply wrong

I'll investigate a little more, but I think we should remove this. I found that every build running in a separate instance, e.g a container or a VM causes this check to send the alarm... not nice

**Comment by schaefi** [comment]
I have disabled the runtime check causing the misleading error

However if code in custom scripts causes additional mounts not done by kiwi itself you still will end up in the situation that filesystems might be busy and stay mounted after kiwi is done. Sometimes programs just mount e.g proc or some sysfs filesystem or initiate a cgroup or send and udev event to the host which causes a mount in the chroot. All this is often hidden to the user but also hidden to kiwi.

If you are building images and stumble over this problem without finding a persistent way to prevent it I recommend to build in a contained environment:

   https://github.com/openSUSE/kiwi/wiki/Building-images-in-a-Docker-container

This will not solve the busy problematic but keep your host system clean and protected

**Comment by vinyanalista** [comment]
I realized that images built with [SUSE Studio](https://susestudio.com/) do not use the `baseMount` and `baseCleanMount` functions, so I commented their calls out from my `config.sh`.

Besides that, I changed the call to the `kiwi` command in the script `create_appliance.sh`:

```sh
run_cmd "$kiwi --build $src/ -d $dst"
```

to manually prepare the image, unmount the chrooted `/sys` and then create the image:

```sh
run_cmd "$kiwi --prepare $src --root $dst/build/image-root"
umount -f -l $dst/build/image-root/sys
run_cmd "$kiwi --create $dst/build/image-root --destdir $dst"
```

So, I borrowed @v-fox suggestion, thank you @v-fox!

Now, I'm able to build images with my VirtualBox virtual machine without issues. I'm using [KIWI v7.03.24](https://github.com/openSUSE/kiwi/tree/v7.03.24) from the official openSUSE Leap 42.1 repository.

Also, thank you @schaefi for your fast answer and bugfix! It is available in the new [KIWI v7.03.35](https://github.com/openSUSE/kiwi/commit/a2146b40dca4b727b46c82d899de260900c327ae), isn't it? Soon I'll give it a try, for now I'm going to keep things that way...
